File: app/src/test/java/org/mifos/mobilebanking/LoanAccountWithdrawPresenterTest.java
package org.mifos.mobilebanking;

import android.content.Context;

import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mifos.mobilebanking.api.DataManager;
import org.mifos.mobilebanking.models.accounts.loan.LoanWithdraw;
import org.mifos.mobilebanking.presenters.LoanAccountWithdrawPresenter;
import org.mifos.mobilebanking.ui.views.LoanAccountWithdrawView;
import org.mifos.mobilebanking.util.RxSchedulersOverrideRule;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import io.reactivex.Observable;
import okhttp3.ResponseBody;

import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

/**
 * Created by dilpreet on 24/7/17.
 */
@RunWith(MockitoJUnitRunner.class)
public class LoanAccountWithdrawPresenterTest {

    @Rule
    public final RxSchedulersOverrideRule mOverrideSchedulersRule = new RxSchedulersOverrideRule();

    @Mock
    Context context;

    @Mock
    DataManager dataManager;

    @Mock
    LoanAccountWithdrawView view;

    @Mock
    LoanWithdraw loanWithdraw;

    @Mock
    ResponseBody mockedResponseBody;

    private LoanAccountWithdrawPresenter presenter;
    private long loanId = 1;

    @Before
    public void setUp() {
        presenter = new LoanAccountWithdrawPresenter(dataManager, context);
        presenter.attachView(view);
    }

    @After
    public void tearDown() {
        presenter.detachView();
    }

    @Test
    public void testWithdrawLoanAccount() {
        when(dataManager.withdrawLoanAccount(loanId, loanWithdraw)).thenReturn(Observable.
                just(mockedResponseBody));

        presenter.withdrawLoanAccount(loanId, loanWithdraw);

        verify(view).showProgress();
        verify(view).hideProgress();
        verify(view).showLoanAccountWithdrawSuccess();
        verify(view, never()).showLoanAccountWithdrawError(context.getString(R.string.
                error_loan_account_withdraw));
    }

    @Test
    public void testWithdrawLoanAccountFail() {
        when(dataManager.withdrawLoanAccount(loanId, loanWithdraw)).thenReturn(Observable.
                <ResponseBody>error(new RuntimeException()));

        presenter.withdrawLoanAccount(loanId, loanWithdraw);

        verify(view).showProgress();
        verify(view).hideProgress();
        verify(view).showLoanAccountWithdrawError(context.getString(R.string.
                error_loan_account_withdraw));
        verify(view, never()).showLoanAccountWithdrawSuccess();
    }

}

File: app/src/test/java/org/mifos/mobile/LoanAccountWithdrawPresenterTest.kt
package org.mifos.mobile

import android.content.Context

import io.reactivex.Observable

import okhttp3.ResponseBody

import org.junit.After
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith

import org.mifos.mobile.api.DataManager
import org.mifos.mobile.models.accounts.loan.LoanWithdraw
import org.mifos.mobile.presenters.LoanAccountWithdrawPresenter
import org.mifos.mobile.ui.views.LoanAccountWithdrawView
import org.mifos.mobile.util.RxSchedulersOverrideRule

import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.junit.MockitoJUnitRunner

/**
 * Created by dilpreet on 24/7/17.
 */
@RunWith(MockitoJUnitRunner::class)
class LoanAccountWithdrawPresenterTest {
    @Rule
    @JvmField
    val mOverrideSchedulersRule = RxSchedulersOverrideRule()

    @Mock
    var context: Context? = null

    @Mock
    var dataManager: DataManager? = null

    @Mock
    var view: LoanAccountWithdrawView? = null

    @Mock
    var loanWithdraw: LoanWithdraw? = null

    @Mock
    var mockedResponseBody: ResponseBody? = null
    private var presenter: LoanAccountWithdrawPresenter? = null
    private val loanId: Long = 1

    @Before
    fun setUp() {
        presenter = LoanAccountWithdrawPresenter(dataManager!!, context)
        presenter?.attachView(view)
    }

    @After
    fun tearDown() {
        presenter?.detachView()
    }

    @Test
    fun testWithdrawLoanAccount() {
        Mockito.`when`(dataManager?.withdrawLoanAccount(loanId, loanWithdraw)).thenReturn(Observable.just(mockedResponseBody))
        presenter?.withdrawLoanAccount(loanId, loanWithdraw)
        Mockito.verify(view)?.showProgress()
        Mockito.verify(view)?.hideProgress()
        Mockito.verify(view)?.showLoanAccountWithdrawSuccess()
        Mockito.verify(view, Mockito.never())?.showLoanAccountWithdrawError(context?.getString(R.string.error_loan_account_withdraw))
    }

    @Test
    fun testWithdrawLoanAccountFail() {
        Mockito.`when`(dataManager?.withdrawLoanAccount(loanId, loanWithdraw)).thenReturn(Observable.error(RuntimeException()))
        presenter?.withdrawLoanAccount(loanId, loanWithdraw)
        Mockito.verify(view)?.showProgress()
        Mockito.verify(view)?.hideProgress()
        Mockito.verify(view)?.showLoanAccountWithdrawError(context?.getString(R.string.error_loan_account_withdraw))
        Mockito.verify(view, Mockito.never())?.showLoanAccountWithdrawSuccess()
    }
}